### PR TITLE
Add FeatureDetector caching.

### PR DESCRIPTION
**Pull request by StoneyJackson**
Implements #36

### CODE REVIEW

File: src/mousetrap/vision.py [hunk 1]
```diff
@@ -87,6 +88,23 @@ class HaarNameError(Exception):
 
 
 class FeatureDetector(object):
+
+    _INSTANCES = {}
+
+    @classmethod
+    def get_detector(cls, config, name, scale_factor=1.1, min_neighbors=3):
+        if name in cls._INSTANCES:
```
Comment (kevin-brown) [inline src/mousetrap/vision.py]: Considering the name is only unique to the haar file used and not the arguments (such as the scale factor or minimum neighbours), we should be storing the combination based on a different key.

Comment (StoneyJackson) [inline src/mousetrap/vision.py]: Good point.

**Comment by StoneyJackson** [comment]
Are we good to merge?

File: src/mousetrap/vision.py [hunk 2]
```diff
@@ -106,15 +126,28 @@ def __init__(self, config, name, scale_factor=1.1, min_neighbors=3):
...
-        self._unpack_first()
-        self._extract_image()
-        self._calculate_center()
-        return self._single
+        if image in self._detect_cache:
+            LOGGER.debug("Detection cache hit: %(image)d -> %(result)s" %
+                    {'image':id(image), 'result':self._detect_cache[image]}
+                    )
+            if isinstance(self._detect_cache[image], FeatureNotFoundException):
+                raise FeatureNotFoundException(str(self._detect_cache[image]))
```
Comment (kevin-brown) [inline src/mousetrap/vision.py]: Is it possible to re-raise the stored exception? While manually re-raising the new exception while using the old `message` works, we will lose any extra data that is stored within the exception (such as the stack trace) and it locks the implementation down quite a bit.

File: src/mousetrap/image.py [hunk 3]
```diff
@@ -36,6 +36,12 @@ def get_width(self):
     def get_height(self):
         return self._image_cv.shape[1]
 
+    def __hash__(self):
+        return id(self)
+
+    def __eq__(self, other):
+        return hash(self) == hash(other)
```
Comment (kevin-brown) [inline src/mousetrap/image.py]: What required us to re-implement both this and the object hash?

**Pull request merged**